linera-storage/src/lib.rs

```diff
@@ -546,3 +546,314 @@ pub trait Clock {
+
+        // Test multiple blob read (read_blobs)
+        let blob_ids = vec![blob_id1, blob_id2, blob_id3];
+        let read_blobs = storage.read_blobs(&blob_ids).await?;
+        assert_eq!(read_blobs.len(), 3);
+
+        // Verify each blob was read correctly
+        assert_eq!(read_blobs[0], Some(test_blob1.clone()));
+        assert_eq!(read_blobs[1], Some(test_blob2));
+        assert_eq!(read_blobs[2], Some(test_blob3));
+
+        // Test missing blobs detection
+        let missing_blob_id = BlobId::new(CryptoHash::test_hash("missing"), BlobType::Data);
+        let missing_blobs = storage.missing_blobs(&[blob_id1, missing_blob_id]).await?;
+        assert_eq!(missing_blobs, vec![missing_blob_id]);
+
+        // Test maybe_write_blobs (should return false as blobs don't have blob states yet)
+        let write_results = storage.maybe_write_blobs(&[test_blob1.clone()]).await?;
+        assert_eq!(write_results, vec![false]);
+
+        // Test blob state operations
+        let blob_state1 = BlobState {
+            last_used_by: None,
+            chain_id: ChainId(CryptoHash::test_hash("chain1")),
+            block_height: BlockHeight(0),
+            epoch: Some(Epoch::ZERO),
+        };
+        let blob_state2 = BlobState {
+            last_used_by: Some(CryptoHash::test_hash("cert")),
+            chain_id: ChainId(CryptoHash::test_hash("chain2")),
+            block_height: BlockHeight(1),
+            epoch: Some(Epoch::from(1)),
+        };
+
+        // Test blob state existence before writing
+        assert!(!storage.contains_blob_state(blob_id1).await?);
+        assert!(!storage.contains_blob_state(blob_id2).await?);
+
+        // Test blob state writing
+        storage
+            .maybe_write_blob_states(&[blob_id1], blob_state1.clone())
+            .await?;
+        storage
+            .maybe_write_blob_states(&[blob_id2], blob_state2.clone())
+            .await?;
+
+        // Test blob state existence after writing
+        assert!(storage.contains_blob_state(blob_id1).await?);
+        assert!(storage.contains_blob_state(blob_id2).await?);
+
+        // Test single blob state read
+        let read_blob_state = storage.read_blob_state(blob_id1).await?;
+        assert_eq!(read_blob_state, Some(blob_state1.clone()));
+
+        // Test multiple blob state read (read_blob_states)
+        let read_blob_states = storage.read_blob_states(&[blob_id1, blob_id2]).await?;
+        assert_eq!(read_blob_states.len(), 2);
+
+        // Verify blob states
+        assert_eq!(read_blob_states[0], Some(blob_state1));
+        assert_eq!(read_blob_states[1], Some(blob_state2));
+
+        // Test maybe_write_blobs now that blob states exist (should return true)
+        let write_results = storage.maybe_write_blobs(&[test_blob1.clone()]).await?;
+        assert_eq!(write_results, vec![true]);
+
+        Ok(())
+    }
+
+    async fn test_storage_certificate<S: Storage + Sync>(storage: &S) -> Result<(), ViewError>
```

Comment:
This only tests non-existent certificates?

Reply:
Ok, corrected.